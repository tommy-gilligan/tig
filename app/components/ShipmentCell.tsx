import React from "react";
import { ShipmentSchema } from "./types/shipment";
import { Box } from "@chakra-ui/react";

export const ShipmentCell = ({
  trackingId,
  lastUpdate,
}: {
  trackingId: ShipmentSchema["trackingId"];
  lastUpdate: ShipmentSchema["lastUpdate"];
}) => {
  const date = new Intl.DateTimeFormat(undefined, {
    day: "numeric",
    month: "short",
    year: "numeric",
  }).format(Date.parse(lastUpdate));

  return (
    <>
      <Box data-testid="tracking-id" fontSize="0.9em">{trackingId}</Box>
      <Box data-testid="last-update" fontSize="0.7em" color="grey">
        Created: {date}
      </Box>
    </>
  );
};
