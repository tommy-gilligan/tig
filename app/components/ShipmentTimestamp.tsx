import React from "react";
import { Text } from "@chakra-ui/react";

export const ShipmentTimestamp = ({ timestamp }: { timestamp: string }) => {
  const time = new Intl.DateTimeFormat(undefined, {
    month: "long",
    year: "numeric",
    day: "numeric",
    hour: "numeric",
    minute: "numeric",
  }).format(Date.parse(timestamp));

  return <Text as="b">{time}</Text>;
};
