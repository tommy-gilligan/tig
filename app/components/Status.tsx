import React from "react";
import { Badge } from "@chakra-ui/react";
import { ShipmentSchema, ShipmentStatus } from "../types/shipment";

export const Status = ({ status }: { status: ShipmentSchema["status"] }) => {
  if (status === ShipmentStatus.DELIVERED) {
    return (
      <Badge
        borderRadius="5px"
        padding="5px"
        paddingLeft="10px"
        paddingRight="10px"
        textTransform="none"
        variant="outline"
        colorScheme="green"
      >
        Delivered
      </Badge>
    );
  } else if (status === ShipmentStatus.IN_TRANSIT) {
    return (
      <Badge
        borderRadius="5px"
        padding="5px"
        paddingLeft="10px"
        paddingRight="10px"
        textTransform="none"
        variant="outline"
      >
        In-Transit
      </Badge>
    );
  } else if (status === ShipmentStatus.MANIFESTED) {
    return (
      <Badge
        borderRadius="5px"
        padding="5px"
        paddingLeft="10px"
        paddingRight="10px"
        textTransform="none"
        variant="outline"
      >
        Manifested
      </Badge>
    );
  } else {
    return (
      <Badge
        borderRadius="5px"
        padding="5px"
        paddingLeft="10px"
        paddingRight="10px"
        textTransform="none"
        variant="outline"
        colorScheme="red"
      >
        Unknown
      </Badge>
    );
  }
};
