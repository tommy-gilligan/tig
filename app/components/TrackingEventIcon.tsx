import React from "react";

import { TrackingEventStatusSeverity } from "../types/tracking_event";

import { ListIcon } from "@chakra-ui/react";
import { CheckCircleIcon, WarningIcon } from "@chakra-ui/icons";

export const TrackingEventIcon = ({
  statusSeverity,
}: {
  statusSeverity: TrackingEventStatusSeverity;
}) => {
  if (statusSeverity === TrackingEventStatusSeverity.SUCCESS) {
    return (
      <ListIcon
        marginLeft="10px"
        boxSize={6}
        as={CheckCircleIcon}
        color="green.500"
      />
    );
  } else if (statusSeverity === TrackingEventStatusSeverity.INFO) {
    return (
      <ListIcon
        marginLeft="10px"
        boxSize={6}
        as={WarningIcon}
        color="red.500"
      />
    );
  } else {
    return (
      <ListIcon
        marginLeft="10px"
        boxSize={6}
        as={WarningIcon}
        color="red.500"
      />
    );
  }
};
