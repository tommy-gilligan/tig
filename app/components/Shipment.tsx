import React from "react";
import { ShipmentSchema } from "../types/shipment";
import { Status } from "./Status";
import { ShipmentTimestamp } from "./ShipmentTimestamp";
import { Table, Tbody, Tr, Td, TableContainer } from "@chakra-ui/react";

export const Shipment = ({ shipment }: { shipment: ShipmentSchema }) => {
  return (
    <TableContainer whiteSpace="wrap">
      <Table variant="unstyled">
        <Tbody>
          <Tr>
            <Td
              padding="0"
              paddingRight="20px"
              paddingBottom="20px"
              color="grey"
              verticalAlign="top"
            >
              Status
            </Td>
            <Td padding="0" paddingBottom="20px" verticalAlign="top">
              <Status status={shipment.status} />
            </Td>
          </Tr>
          {shipment.deliveredTime && (
            <Tr>
              <Td
                padding="0"
                paddingRight="20px"
                paddingBottom="20px"
                color="grey"
                verticalAlign="top"
              >
                Delivered time
              </Td>
              <Td
                data-testid="delivered-time"
                padding="0"
                paddingBottom="20px"
                verticalAlign="top"
              >
                <ShipmentTimestamp timestamp={shipment.deliveredTime} />
              </Td>
            </Tr>
          )}
          <Tr>
            <Td
              padding="0"
              paddingRight="20px"
              paddingBottom="20px"
              color="grey"
              verticalAlign="top"
            >
              Delivery address
            </Td>
            <Td padding="0" paddingBottom="20px" verticalAlign="top">
              {shipment.deliveryAddress}
            </Td>
          </Tr>
          <Tr>
            <Td
              padding="0"
              paddingRight="20px"
              paddingBottom="20px"
              color="grey"
              verticalAlign="top"
            >
              Last updated
            </Td>
            <Td padding="0" paddingBottom="20px" verticalAlign="top">
              <ShipmentTimestamp timestamp={shipment.lastUpdate} />
            </Td>
          </Tr>
          <Tr>
            <Td
              padding="0"
              paddingRight="20px"
              paddingBottom="20px"
              color="grey"
              verticalAlign="top"
            >
              Total transit time
            </Td>
            <Td padding="0" paddingBottom="20px" verticalAlign="top">
              {shipment.totalTransit}
            </Td>
          </Tr>
        </Tbody>
      </Table>
    </TableContainer>
  );
};
