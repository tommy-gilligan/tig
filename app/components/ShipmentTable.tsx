import React from "react";
import {
  Flex,
  Table,
  Thead,
  Tbody,
  Td,
  Tr,
  Th,
  TableContainer,
} from "@chakra-ui/react";
import { ShipmentSchema } from "./types/shipment";
import { Status } from "./Status";
import { ShipmentCell } from "./ShipmentCell";
import {
  createColumnHelper,
  flexRender,
  getCoreRowModel,
  getSortedRowModel,
  useReactTable,
  SortingState,
  Row,
  Column,
} from "@tanstack/react-table";
import { ArrowUpIcon, ArrowDownIcon, ArrowUpDownIcon } from "@chakra-ui/icons";

const shipmentSort = (
  rowA: Row<ShipmentSchema>,
  rowB: Row<ShipmentSchema>,
  _: string,
): number => {
  const lastUpdateA = rowA.original.lastUpdate;
  const lastUpdateB = rowB.original.lastUpdate;

  if (lastUpdateA < lastUpdateB) {
    return -1;
  } else if (lastUpdateA > lastUpdateB) {
    return 1;
  }
  return 0;
};

const columnHelper = createColumnHelper<ShipmentSchema>();
const columns = [
  columnHelper.accessor(
    (row) => ({ trackingId: row.trackingId, lastUpdate: row.lastUpdate }),
    {
      id: "shipment",
      header: "Shipment",
      cell: (info) => {
        const value = info.getValue();
        const trackingId = value.trackingId;
        const lastUpdate = value.lastUpdate;
        return <ShipmentCell trackingId={trackingId} lastUpdate={lastUpdate} />;
      },
      sortingFn: shipmentSort,
    },
  ),
  columnHelper.accessor((row) => row.status, {
    header: "Status",
    id: "status",
    cell: (info) => <Status status={info.getValue()} />,
  }),
];

const SortIcon = ({ column }: { column: Column<ShipmentSchema> }) => {
  if (column.getCanSort()) {
    if ((column.getIsSorted() as string) === "asc") {
      return <ArrowDownIcon marginLeft="10px" />;
    } else if ((column.getIsSorted() as string) === "desc") {
      return <ArrowUpIcon marginLeft="10px" />;
    } else {
      return <ArrowUpDownIcon marginLeft="10px" />;
    }
  } else {
    return null;
  }
};

export const ShipmentTable = ({
  shipments,
  onShipmentSelected,
}: {
  shipments: [ShipmentSchema];
  onShipmentSelected: (value: ShipmentSchema | undefined) => void;
}) => {
  const [sorting, setSorting] = React.useState<SortingState>([]);
  const table = useReactTable({
    data: shipments,
    columns,
    state: {
      sorting,
    },
    onSortingChange: setSorting,
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
  });

  return (
    <TableContainer background="white" borderRadius="5px">
      <Table variant="simple">
        <Thead>
          {table.getHeaderGroups().map((headerGroup) => (
            <Tr key={headerGroup.id}>
              {headerGroup.headers.map((header) => (
                <Th
                  padding="0"
                  fontWeight="normal"
                  textTransform="none"
                  key={header.id}
                >
                  <Flex
                    padding="13px 20px"
                    align="center"
                    onClick={header.column.getToggleSortingHandler()}
                    {...{ cursor: header.column.getCanSort() ? "pointer" : "" }}
                  >
                    {flexRender(
                      header.column.columnDef.header,
                      header.getContext(),
                    )}
                    <SortIcon column={header.column} />
                  </Flex>
                </Th>
              ))}
            </Tr>
          ))}
        </Thead>
        <Tbody>
          {table.getRowModel().rows.map((row) => {
            return (
              <Tr
                cursor="pointer"
                onClick={() => onShipmentSelected(row.original)}
                key={row.id}
		data-testid="shipment-row"
              >
                {row.getVisibleCells().map((cell) => (
                  <Td padding="13px 20px" verticalAlign="center" key={cell.id}>
                    {flexRender(cell.column.columnDef.cell, cell.getContext())}
                  </Td>
                ))}
              </Tr>
            );
          })}
        </Tbody>
      </Table>
    </TableContainer>
  );
}
