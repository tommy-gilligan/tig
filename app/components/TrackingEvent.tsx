import React from "react";
import { TrackingEventSchema } from "./types/tracking_event";
import { Text, Flex, Box, List, ListItem, ListIcon } from "@chakra-ui/react";
import { gql, useQuery } from "@apollo/client";
import { TrackingEventTimestamp } from "./TrackingEventTimestamp"
import { TrackingEventIcon } from "./TrackingEventIcon"

export const TrackingEvent = ({
  trackingEvent: {statusSeverity, status, location, timestamp},
}: {
  trackingEvent: TrackingEventSchema;
}) => {
  return (
    <ListItem>
      <Flex align="center">
        <TrackingEventIcon statusSeverity={statusSeverity} />

        <Flex
          borderBottomColor="grey"
          borderBottomStyle="dashed"
          borderBottomWidth="1px"
          paddingTop="10px"
          paddingBottom="10px"
          justify="space-between"
          width="100%"
          marginRight="10px"
        >
          <Box>
            <Text data-testid="status" as="b">{status}</Text>
            <br />
            <Text data-testid="location">{location}</Text>
          </Box>

          <TrackingEventTimestamp timestamp={timestamp} />
        </Flex>
      </Flex>
    </ListItem>
  );
};
