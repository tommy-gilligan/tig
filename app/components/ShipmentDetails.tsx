import React from "react";
import { ShipmentSchema } from "./types/shipment";
import {
  Drawer,
  DrawerBody,
  DrawerHeader,
  DrawerOverlay,
  DrawerContent,
  DrawerCloseButton,
  Accordion,
  AccordionItem,
  AccordionButton,
  AccordionPanel,
  AccordionIcon,
  Box,
  Flex,
} from "@chakra-ui/react";
import { TrackingHistory } from "./TrackingHistory";
import { Shipment } from "./Shipment";

export const ShipmentDetails = ({
  shipment,
  onClose,
}: {
  shipment?: ShipmentSchema;
  onClose: () => void;
}) => {
  const isOpen = !!shipment;

  return (
    <Drawer size="lg" isOpen={isOpen} placement="right" onClose={onClose}>
      <DrawerOverlay />
      <DrawerContent>
        <DrawerHeader padding="15px">
          {isOpen && shipment.trackingId}
        </DrawerHeader>
        <DrawerCloseButton />
        <DrawerBody padding="0">
          <Accordion allowMultiple defaultIndex={[0, 1]}>
            <AccordionItem>
              <AccordionButton color="grey" fontWeight="bold">
                <Flex justify="space-between" width="100%">
                  <Box>SHIPMENT</Box>
                  <AccordionIcon />
                </Flex>
              </AccordionButton>
              <AccordionPanel>
                {isOpen && <Shipment shipment={shipment} />}
              </AccordionPanel>
            </AccordionItem>
            <AccordionItem>
              <AccordionButton color="grey" fontWeight="bold">
                <Flex justify="space-between" width="100%">
                  <Box>TRACKING HISTORY</Box>
                  <AccordionIcon />
                </Flex>
              </AccordionButton>
              <AccordionPanel>
                {isOpen && <TrackingHistory trackingId={shipment.trackingId} />}
              </AccordionPanel>
            </AccordionItem>
          </Accordion>
        </DrawerBody>
      </DrawerContent>
    </Drawer>
  );
}
