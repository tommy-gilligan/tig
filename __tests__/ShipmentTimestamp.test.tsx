import { screen, render } from '@testing-library/react'
import { ShipmentTimestamp } from '../app/components/ShipmentTimestamp'

it('renders a shipment timestamp', () => {
  const { container } = render(
    <ShipmentTimestamp timestamp="2024-01-03T02:23:45Z" />
  )
  expect(container).toMatchSnapshot()
})
