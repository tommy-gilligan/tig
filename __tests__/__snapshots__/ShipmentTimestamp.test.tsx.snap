// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a shipment timestamp 1`] = `
<div>
  <b
    class="chakra-text css-0"
  >
    January 3, 2024 at 1:23 PM
  </b>
</div>
`;
